### ISSUE TITLE
The gradient of jax.lax.select gives nan when the false branch is nan

### ISSUE THREAD

**Issue by kkew3** [google/jax#20601]
### Description

A more verbose version of the issue: the false branch somehow affects the gradient in an erroneous way. In contrast, `jax.lax.cond` has no such issue. However, `cond` is converted to `select` when `vmap`- or `pmap`-ed ...

I found issue #347 that may be related to this issue, but it was closed, while that issue still persists in my `jax` installation.

Minimal reproducible example:

```python
import jax
from jax import lax
from jax import numpy as jnp

def g_select(beta, x):
    # when x is nan, beta * x will be nan
    return lax.select(False, beta * x, 1.0)

g_select_fun_and_grad = jax.value_and_grad(g_select)

def main():
    beta = jnp.array(2.3)
    print(g_select_fun_and_grad(beta, jnp.nan))  # -> (1.0, nan)
    print(g_select_fun_and_grad(beta, -2.0))  # -> (1.0, -0.0)
    print(g_select_fun_and_grad(beta, 2.0))  # -> (1.0, 0.0)

if __name__ == '__main__':
    main()
```

### System info (python version, jaxlib version, accelerator, etc.)

This issue is observed on two hosts:

```
jax:    0.4.25
jaxlib: 0.4.25
numpy:  1.26.4
python: 3.9.7 (v3.9.7:1016ef3790, Aug 30 2021, 16:39:15)  [Clang 6.0 (clang-600.0.57)]
jax.devices (1 total, 1 local): [CpuDevice(id=0)]
process_count: 1
platform: uname_result(system='Darwin', node='kw-Mac', release='21.6.0', version='Darwin Kernel Version 21.6.0: Mon Feb 19 20:24:34 PST 2024; root:xnu-8020.240.18.707.4~1/RELEASE_X86_64', machine='x86_64')
```

```
jax:    0.4.25
jaxlib: 0.4.25
numpy:  1.26.4
python: 3.11.5 (main, Sep 11 2023, 13:54:46) [GCC 11.2.0]
jax.devices (1 total, 1 local): [CpuDevice(id=0)]
process_count: 1
platform: uname_result(system='Linux', node='dell-p3460', release='5.15.0-101-generic', version='#111-Ubuntu SMP Tue Mar 5 20:16:58 UTC 2024', machine='x86_64')
```

**Comment by kkew3** [comment]
It turns out that the [doc](https://jax.readthedocs.io/en/latest/faq.html#gradients-contain-nan-where-using-where) already mentions this. In my use case, I use nan to pad variable length input so that vmap can be applied on them, which should be masked out eventually. Such choice is made so that whenever the masking fails, the result should become nan, and I'll be aware of that.

Now I learn that this won't work. Sorry for the interruption. But could anyone suggest an alternative approach to tackle my use case? Like how to correctly pad and mask variable length input? Thank you so much!

**Comment by jakevdp** [comment]
I'd suggest padding with a finite but invalid value.
